### ISSUE TITLE
Create Guides Markdown Converter (Google Apps Script)

### ISSUE THREAD

**Issue by abenipa3** [hackforla/website#2978]
*****************************

## ==> SEE [HISTORY WIKI](https://github.com/hackforla/knowledgebase/wiki/History) FOR WORK DONE SO FAR <==

*****************************
### Overview
As developers, we want a Google Apps Script to convert the current guides in Google Docs to markdown files, so that the markdown files can be used to display the guides on our website. 

#### List of Features Required for the MD Converter
The following are features of what we need the MD Converter to do:
-	Find the file (guide) within the folder or drive. 
o	File can be found by reference or name if it exists.
-	Reads the contents of the file, including:
o	Headings
o	Bulleted lists
o	Numbered lists
o	Multileveled lists
o	Indentations and Spaces
o	Font styles (bold, italics, underline)
o	URLS
o	Images
o	Text blocks, or areas of code nested inside of the document.
-	Formats the contents of the file accordingly.
-	Returns image paths where images were specifically inserted in the document.
-	Creates a folder as an output with the following contents:
o	File as a Markdown, generated by name. (Example: “how-to-set-reminders-in-slack.md”)
o	Creates a folder to save images that were inserted within the Google Doc.

<details> <summary>Front Matter</summary>

```
layout: guide-pages
title: How to Set Reminders in Slack
provider-link: "/how-to-set-reminders-in-slack"
overview: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum." 
guide-author:
  - name: "Jane Doe"
    links:
      linked-in: "https://www.linkedin.com/in/jane-doe/"
      github: "https://github.com/jane-doe"
    picture: https://avatars.githubusercontent.com/jane-doe
```

</details>

### History
  - Alyssa built the Apps Script as seen in this [link](https://script.google.com/home/projects/1CFED97nWJO_zVzDvrN6wUztJVZQaDFUhrSKe3BXaaHhV02X5AgqT4mRR/edit). Alyssa is moving off the project and @ethanstrominger is now working on it.
  - [GoogleDoc2Html](https://github.com/oazabir/GoogleDoc2Html) is an open source project that was used as a base in July 2022 by Ethan Strominger. It was the top hit in google. However, searching npmjs.com turned out to be much more fruitful.

### Resources
- List of Tools can be found on [Admin wiki Tools](https://github.com/hackforla/admin/wiki/tools)
- List of Program Areas can be found on [hack for LA website projects page ](https://www.hackforla.org/projects/)

**Comment by JessicaLucindaCheng** [comment]
Next steps:
- [x] Discuss this issue with Bonnie.
- [x] Once we move the Google Apps Script and any relevant file(s) to [Markdown Converter folder](https://drive.google.com/drive/folders/1A7fCpYTakb_3wCOv7fSM8oyN-wTkisA0), the following links need to be updated:
  - [x] Action Items section
    > Alyssa has started building the Apps Script as seen in this [link](https://script.google.com/d/1iNnRzYQW-6a24ScyivlIb1yEgB7Jloshct5R_LUvzTBz6jcX_IDtX_TN/edit?usp=sharing). (Link subject to change once we determine where to move the file in HfLA folder)
  
  - [x] Resources/Instructions section
    > - [Current Apps Script](https://script.google.com/d/1iNnRzYQW-6a24ScyivlIb1yEgB7Jloshct5R_LUvzTBz6jcX_IDtX_TN/edit?usp=sharing) - Subject to change once we determine where to move file in HfLA folder

**Comment by JessicaLucindaCheng** [comment]
@alyssabenipayo 
- [x] Include a list of features. In other words, what exactly do we need the md converter to do.
- [x] Identify existing converter in the issue that is currently used.

@JessicaLucindaCheng 
- [x] Add an action item to identify additional converters that might work.
- [x] Add an action item to identify/look for if there are any open-source code that we can use in our own converter by crediting them. This will be dependent on the source code license and decide if this is allowed.
- [x] Add an action item to do a C&C analysis on it.
- [x] Make a copy of this C&C: https://docs.google.com/spreadsheets/d/1ePxxsLpdC4MMvJICyMYljHawhs1j-hpFHZFOtB0fdDc/edit#gid=1259306930
- [x] add a link to the copy of the C&C to this issue.
- [x] Move Markdown Converter folder to Feature Branches folder

**Comment by abenipa3** [comment]
# List of Features Required
The following are features of what we need the MD Converter to do:
-	Find the file (guide) within the folder or drive. 
o	File can be found by reference or name if it exists.
-	Reads the contents of the file, including:
o	Headings
o	Bulleted lists
o	Numbered lists
o	Multileveled lists
o	Indentations and Spaces
o	Font styles (bold, italics, underline)
o	URLS
o	Images
o	Text blocks, or areas of code nested inside of the document.
-	Formats the contents of the file accordingly.
-	Returns image paths where images were specifically inserted in the document.
-	Creates a folder as an output with the following contents:
o	File as a Markdown, generated by name. (Example: “how-to-set-reminders-in-slack.md”)
o	Creates a folder to save images that were inserted within the Google Doc.

# Existing Converter
## Docs to Markdown
-	Docs to Markdown - Free, open-source Drive add-on that converts a Google Doc to simple, readable Markdown or HTML. (Available in Google Workspace Marketplace)
-	Link to Extension: https://workspace.google.com/marketplace/app/docs_to_markdown/700168918607
- The next sections are also mentioned in this issue: https://github.com/hackforla/guides/issues/10#issuecomment-1066190743

### Summary: 
As of January 8, 2022, the Docs to Markdown Converter released an update that removes error/warning messages called `reckless mode`. 

However, a message as seen below still appears at the top of the converted MD file and there are more issues encountered (mentioned in the next section) after converting the document. 
![image](https://user-images.githubusercontent.com/38295612/158080504-07773395-43b1-4427-9f2e-c69883a4723c.png)
### New Issues Encountered

<details> <summary> Images are not in the correct order </summary> 

Expected Output 
![right-order](https://user-images.githubusercontent.com/38295612/158072463-0497a1f4-88e7-4fe5-a22f-80eeed20f60c.png)

Actual Output
![wrong-order](https://user-images.githubusercontent.com/38295612/158072457-1fd97035-a67c-41a3-a221-780b91fa8edf.png)

</details>

<details> <summary> Backslashes appear randomly in the file </summary> 

![Screenshot 2022-03-12 174359](https://user-images.githubusercontent.com/38295612/158070915-a6937dde-5b84-46cc-b697-225b2b14021b.png)

</details>

**Comment by ethanstrominger** [comment]
Progress: Alyssa was working on this and I think I am now the only developer working on it.  
- Reviewed issue description and followed all links.  
- Ran script developed by Alyssa
- Tried out Google Docs to HTML converter (see findings)
- Planned out next steps 
- Modified this issue description with actions.  Some of the actions are questions.

Blockers: None
Availability: Next week 8 hours.
ETA: 4 - 8 weeks?  Will have better sense at end of next week.

**Comment by ethanstrominger** [comment]
Progress: Created a script that recursively converts google docs to html and saves the html on to google drive.  Source and target directories are currently hardcoded.  I am currently storing in a repository on my github and will be meeting with Tamara Snyder, tech lead, to figure out where to store on HackforLA.  Repsitory is [here](https://github.com/ethanstrominger/html-markdown-googledocs-conversion).

**Comment by JessicaLucindaCheng** [comment]
@ethanstrominger Since you are working on this issue, please move it to the "In Progress" column for the "Project Board". You can do this by clicking on the dropdown arrow (as circled below) and selecting "In Progress". Thanks.

<img src="https://user-images.githubusercontent.com/31293603/175827919-845d4e6e-a9bc-48b6-a093-be892399a3dd.png">

**Comment by ethanstrominger** [comment]
.

**Comment by ethanstrominger** [comment]
ARCHIVED FOLLOWING TASKS AS NO LONGER APPLICABLE:
ARCHIVE: New approach - see 
- [] Call from node script  and write to directory
- [] Create subdirectories that match phase
- [] Identify where to store Google Docs used for testing.
- [] Document process for importing and exporting from Google Apps Script and doing source control
- [] Create automated test using "Gold Data" created when a conversion for a Google doc is deemed successful.   
- [] Create samples to convert
- [] Write to directory within web site
- [] Document process for importing and exporting from Google Apps Script and doing source control
- [] Create automated test using "Gold Data" created when a conversion for a Google doc is deemed successful.
- [] Create express server and route for kicking off job
- [] Make it possible to convert from a shared drive (currently, only user's folders can be accessed)
- [] Try existing manuals
- [] Add support for images
- [] Add support for numbered lists with paragraphs interspersed between items (requires start with number)
- [] Create a method that replaces every double quote with backticks in the markdown, e.g. Enter "/Remind" into the field, into copy and paste text, e.g., Enter `/Remind` into the field. More information [here](https://github.com/hackforla/guides/issues/10#issuecomment-936572293).
- [] Write up proposal for how this will work in proposal.md
- [] Figure out and write up how to run using clasp from terminal.
- [] Correct numbering for when lists split
- [] Figure out where to put documents for converting
- [] Create a method that replaces every double quote with backticks in the markdown, e.g. Enter "/Remind" into the field, into copy and paste text, e.g., Enter `/Remind` into the field. More information [here](https://github.com/hackforla/guides/issues/10#issuecomment-936572293).
- [] Create a method that would automatically set the full image path for images in generated image folder.
  - The current method returns `![image alt text](/image-1.png)`
  - Expected output: `![[image alt text](../assets/images/guides/how-to-set-reminders-in-slack/image2.png "image_tooltip")`
- [] Create a method that will automatically generate the front matter as seen below:
- [] Once the pull request associated with this issue is approved and merged, please update and edit issue #1515 by
  - [] Checking off the dependency for this issue
  - [] If all dependencies are checked off, please move issue #1515 to the New Issue Approval column and remove the Dependency label

Alyssa built the Apps Script as seen in this [link](https://script.google.com/home/projects/1CFED97nWJO_zVzDvrN6wUztJVZQaDFUhrSKe3BXaaHhV02X5AgqT4mRR/edit). Alyssa is moving off the project and @ethanstrominger is now working on it.  

For testing purposes, "How to Set Reminders in Slack" was used in development of that script. 

Work left to be done on the script (note: may use an existing open source project - see findings below.


Initial steps for Ethan (completed):

- [X] Review work done by Alyssa documented in this issue
- [X] Investigate existing open source (see findings below)
- [X] Check with Alyssa if any additional resources created and anything checked into GitHub
- [X] Find out where to store collateral: 
   - [X] Is separate repository an option?  
   - [X] If existing repository, where does the collateral get stored
- [X] Find out if converting to HTML is acceptable.  There is an open source project that already does this (see Findings) that does a great job.  
consist of google docs to be converted and HTML or markdown equivalent. 
- [X] Create sample Google docs with one or two features to be tested and one document with all the features.
- [x] Investigate how to import and export from Google Apps Script.  The scripts have to be written and run in Google Apps Script.  Import/export is needed to be able to check in to github.
- [x] Investigate existing markdown to Google Docs apps.  This will help in creating scripts.
- [ x] Create initial script either based on Alyssa's script or an open source project for converting a single document to markdown.
- [x] Modify script to convert all documents in a specified directory

**Comment by ethanstrominger** [comment]
Progress: Archived previous repository and started new repository https://github.com/ethanstrominger/hfla-googledocs-converter.   Previous repository could files in a folder and subfolders in Google Docs and convert to html.  However, this is now being abandoned as an already completed npmjs package.  New repository has this package installed without any other changes as a starting point.    
Blockers: None.
Availability: 10 hours
Plan: Create node program that calls npmjs package and document how to  use.
ETA: End of August

**Comment by ethanstrominger** [comment]
Blockers: See bottom of issue description for info on blockers.

**Comment by ethanstrominger** [comment]
Progress: Created a document with info about requirements, deployment, "jekyllifying", how to package for npmjs, and more.  Info needed: Bonnie thought Alyssa had made code that converted How to Set Reminders in Slack to a version that looks good on the website and replicates screenshot on figma.  I sent Alyssa a slack
Blockers: None.
Availability: 0 this week (on vacation) 8 the following week
Plan: When I get back review Slack from Alyssa and potentially adjust my approach, otherwise continue with items on my To Do list.
ETA: Mid Sep I will either have a working version or have enough info to more accurately predict.

**Comment by abenipa3** [comment]
***** SEE https://docs.google.com/document/d/1Tx17ewpeI8r7u_-nSazJtUPjKsFeiSYWu-qmJAsbV3I/edit# *****

@ethanstrominger Thanks for reaching to me with this! Posting the following steps here for our team's reference. 

Added this comment/documentation under this issue's Resources/Instructions as well.

### Here are the steps to how we produce the MD file with the Apps Script:
- Must access to HfLA Google Drive to view the [Google Docs to Markdown Converter folder](https://drive.google.com/drive/folders/1A7fCpYTakb_3wCOv7fSM8oyN-wTkisA0).
- Click on the `GuidesMarkdownConverter - HfLA`. In the first file (`Main.gs`), we see the following:
![image](https://user-images.githubusercontent.com/38295612/185718526-3bba8aa4-663b-4aab-8791-c671458f5e27.png)
- Two important parts of the function to keep in mind are:
  - `var file = findFileByName("How to Set Reminders in Slack");`: For testing purposes (at least in the meantime), let's not change the name here. This locates the file within the folder by name, which in this case is "How to Set Reminders in Slack." 
![image](https://user-images.githubusercontent.com/38295612/185720961-f5b155d8-55b0-488f-bc07-416e2e6edd42.png)
  - ` var newFolder = getFolder("MDConverted", true);` : After running the command, it creates a new folder called `MDConverted` in the ROOT directory ("My Drive"/your personal drive).
- To produce, we click on `Run`. Note that we need to review permissions to do so.
![image](https://user-images.githubusercontent.com/38295612/185721231-86464d5b-a3c5-4b3d-8041-50a1426d5cda.png)
- Click on "Review Permissions" and proceed with granting the script permission to view your drive.
- At the bottom of the screen, we see the Execution Log, which will tell us when it starts and finishes generating the files.
![image](https://user-images.githubusercontent.com/38295612/185721612-9e400246-1fa4-429f-b04b-cc4fd03aa5a5.png)
- Then, head to "My Drive" and locate a new folder called "MDConverted"
![image](https://user-images.githubusercontent.com/38295612/185721385-bb1494ba-8e29-4cdb-bd82-3067afd08c18.png)
- Click on the folder, and then click on the latest "Output" Folder.
![image](https://user-images.githubusercontent.com/38295612/185721634-86493f66-0f0e-426d-ac28-9cdbb7b4671f.png)
  - Side Note: Feel free to change the name of the "Output" folder in line 14 of the script here: ` var subFolder = getFolder("Output-"+getTimeStamp(),true, newFolder);` 
- Within the "Output" Folder, we have the following contents:
  - The MD file for "How to Set Reminders in Slack".
  - A new folder that contains all of the downloaded images from the original Google Doc.
![image](https://user-images.githubusercontent.com/38295612/185721793-59359bd7-0b8b-4b48-861a-3afb90336ae1.png)
- The next thing to do from here is download the "Output" folder.
![image](https://user-images.githubusercontent.com/38295612/185721904-67932650-660f-4b7a-aa8a-c1002088fa9c.png)
- Locate the downloaded "Output" zipped folder and extract it.
- Move the newly MD-converted file into the `website` folder (or whichever directory needed to view/edit the file)
![image](https://user-images.githubusercontent.com/38295612/185722103-3cc19f66-4bcc-492f-b675-bb59bd02c0de.png)
- Move the `images` folder into the directory.
- Jekyll will automatically convert the new MD file into HTML as long as there's a Front Matter as seen below. 
![image](https://user-images.githubusercontent.com/38295612/185722180-30608031-5faf-416b-82c6-d5170d22ed98.png)
- As a result, the output will look like this:
![image](https://user-images.githubusercontent.com/38295612/185723669-021d3936-da01-41d9-a626-c28c55693eab.png)

## Additional Notes - TOC

- The `_includes\toc.html`  generates the section and subsection titles for the sticky navigation of the Guide Pages.
  - This is the same method used for 100Automations' Guide Pages as seen [[here]](https://github.com/100Automations/Website/blob/gh-pages/_includes/toc.html). 
- Additional Source: [[100 Automation All Guides - Live Site]](https://100automations.org/all_guides).

## Caveats

As mentioned earlier, the MD file will automatically be converted into HTML. 

However, we had to input/edit the following manual:

1) Write out the Front Matter.
    - This will automatically convert the MD to HTML.
2) Arrange the images in the right order.
3) Edit the image paths so that the MD file will grab the appropriate image from its respective directory.

Please see images below.

<details> <summary>Images - Current Output</summary>

![image](https://user-images.githubusercontent.com/38295612/185723910-c06232d7-b89b-4454-8d90-a03316f8ae97.png)
![image](https://user-images.githubusercontent.com/38295612/185723799-142c1644-fbbb-4806-b677-28f7967e3432.png)

</details>

<details> <summary>Images - Expected Output</summary>

![image](https://user-images.githubusercontent.com/38295612/185723974-e08cf3c1-4e6d-4cea-97f4-7dd21fd52f49.png)
![image](https://user-images.githubusercontent.com/38295612/185723947-87a663f3-2e17-4b22-a65a-ac1d0926a302.png)

</details>

## Previous Action Items

It seems a lot of changes were made to this issue's description, but here were the Action Items to continue building this App Script:

- [ ] Create a method that will search and return all bulleted and numbered lists in MD syntax.
- [ ] Update the following variable `var file = findFileByName("How to Set Reminders in Slack")`; in `Main.gs` so that it acts as an add-on for any open documents rather than the name of the file.
- [ ] The current method only searches for the file name. We want to use this script as an add-on so we can apply it to any guide.
- [ ] For extra references, ask for permission from HfLA Website Team to access the Apps Script for Wins to Review folder.
- [ ] Create a method that replaces every double quote with backticks in the markdown, e.g. Enter "/Remind" into the field, into copy and paste text, e.g., Enter /Remind into the field. More information [here](https://github.com/hackforla/guides/issues/10#issuecomment-936572293).
- [ ] Create a method that would automatically set the full image path for images in generated image folder.
  - The current method returns`![image alt text](/image-1.png)`
  - Expected output: `![[image alt text](../assets/images/guides/how-to-set-reminders-in-slack/image2.png "image_tooltip")`
- [ ] Create a method that will automatically generate the front matter as seen below:
<details> <summary>Front Matter </summary>

```
layout: guide-pages
title: How to Set Reminders in Slack
provider-link: "/how-to-set-reminders-in-slack"
overview: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum." 
guide-author:
  - name: "Jane Doe"
    links:
      linked-in: "https://www.linkedin.com/in/jane-doe/"
      github: "https://github.com/jane-doe"
    picture: https://avatars.githubusercontent.com/jane-doe
```

</details>


Please feel free to let me know if you have any further questions. Thanks again for taking on this issue.
